강원대 FE 유다연 - 3단계 단체 주문 기능 구현하기

daaoooy opened the pull request:
### 3단계 구현 사항
- react-hook-form으로 주문하기 폼 리팩터링
- 받는 사람 기능 고도화 
    -> 모달을 사용하여 받는 사람 추가할 수 있도록 하기 - Input, InputList, 모달 컴포넌트 따로 구현
    -> 여러 명 받는 사람 추가 가능하도록 하기 (최대 10명) - 버튼 비활성화 사용 
    -> 전화번호 중복 X -> react-hook-form 에서 validate 사용

---

### 피드백 반영
- useFormContext를 사용하여 prop drilling 줄이기
- zod schema를 사용하여 각 필드의 validation 조건, 메세지 등을 응집도 있게 관리하기

daaoooy committed:
docs(readme) : 3단계 과제 요구사항 작성

daaoooy committed:
chore : react-hook-form, zod 설치

daaoooy committed:
feat(button) : secondary 버튼 비활성화 상태 추가

daaoooy committed:
feat(order) : react-hook-form 사용하여 리팩터링

daaoooy committed:
feat(receiver) : 받는 사람 기능 고도화(단체 주문 기능)

- 받는 사람을 여러 명 (최대 10명) 등록할 수 있는 기능을 추가함
- 10명의 정보가 정확하게 입력되지 않으면 정보 반영 되지 않는 기능 추가함
- 받는 사람들끼리는 전화번호가 중복될 수 없는 기능 추가함
- 그 외 조건은 이전과 동일함

- ReceiverInput -> 받는 사람 입력 폼
- ReceiverInputList -> 받는 사람들 쭉 보여주는 리스트
- ReceiverModal -> 받는 사람을 추가하는 모달창
- ReceiverSection -> 최종 반영하여 주문하기 폼에 보여주는 받는 사람 부분

yujo11 commented on src/features/Order/pages/OrderPage.tsx:
```diff
@@ -1,86 +1,80 @@
-        message={order.message}
-        setMessage={(msg) => setOrder({ message: msg })}
-        error={errors.message}
+        register={register('message', { required: '메시지를 입력해주세요.' })}
```
`zod`가 pakcage에 추가되어있는데 zod를 사용해서 schema를 정의해서 사용해보시면 좋을거 같아요

daaoooy replied:
수정 완료

yujo11 commented on src/features/Order/pages/OrderPage.tsx:
```diff
@@ -1,86 +1,80 @@
-          quantity: errors.quantity || '',
-        }}
+
+      <ReceiverSection
```
`useFormContext`라는 react-hook-form에서 제공해주는 hook을 찾아서 적용해보셔도 좋을거 같네요

daaoooy replied:
수정 완료

yujo11 reviewed (changes requested):
안녕하세요 다연님 전체적으로 깔끔하게 잘 구현해주신거 같아요!

전체적으로 아래 2가지를 신경쓰면서 개선을 해보시면 좋을거 같습니다!

1. useFormContext를 사용해서 prop 드릴링을 줄이고
2. zod의 schema를 사용해서 각 필드의 validation 조건, 메세지 등을 응집도 있게 관리

고생 많으셨습니다 ~~ 👍 👍 👍

daaoooy committed:
chore : @hookform/resolvers 설치

daaoooy committed:
refactor : zod, useFormContext 반영

- zod 기반 유효성 검증 구현
- useFormContext 적용하여 prop drilling 개선
- 받는 사람이 없을 시, "받는 사람은 최소 1명이어야 합니다" 에러 뜨도록 개선
- 모달에서 받는 사람을 추가할 때 마다 BottomPurchaseBar에 가격이
  바로바로 반영 되지 않고 모달의 확인 버튼을 눌렀을 때 가격이 반영되도록 변경

yujo11 commented on src/features/Order/components/MessageInput/MessageInput.tsx:
```diff
@@ -1,20 +1,23 @@
+  const {
+    register,
+    formState: { errors },
+  } = useFormContext<Order>()
```
👍 👍 👍

yujo11 commented on src/features/Order/schema/orderSchema.ts:
```diff
@@ -0,0 +1,25 @@
+  quantity: z.number().min(1, '수량은 최소 1개 이상이어야 합니다.'),
+})
+
+export const orderSchema = z.object({
```
👍 👍 👍

yujo11 reviewed (approved):
안녕하세요 다연님 ~

사용이 익숙하지 않았을 수도 있었을거 같은데 잘 사용해주셨네요 ㅎㅎ 

미션 진행하느라 고생 많으셨습니다 ~~ 👍 👍 👍
